File: scala_demo/src/com/gy/wc/SparkWordCount.scala
package com.gy.wc

import org.apache.spark.{SparkConf, SparkContext}

object SparkWordCount {

  val fileName = "./scala_demo/words"

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setMaster("local").setAppName("aparkWc")
    val sc = new SparkContext(conf)
    //    simpleWC(sc)
    commonWC(sc)
    sc.stop()

  }


  def simpleWC(sc: SparkContext): Unit = {
    sc.textFile(fileName).flatMap(_.split(" ")).map((_, 1)).reduceByKey(_ + _).sortBy(_._2, false).foreach(println)
  }

  /**
    * 普通wc
    */
  def commonWC(sc: SparkContext): Unit = {
    val lines = sc.textFile(fileName)
    val words = lines.flatMap(x => x.split(" "))
    val pairWords = words.map(x => (x, 1))
    val result = pairWords.reduceByKey((x: Int, y: Int) => x + y)
    val sort = result.map(_.swap).sortByKey(false).map(x => (x._2, x._1))
    sort.foreach(println)

  }


}

File: Spark_demo/src/com/gy/wc/SparkWordCount.java
package com.gy.wc;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

import java.util.Arrays;

public class SparkWordCount {

    static String fileName = "./scala_demo/words";

    public static void main(String[] args) {

        SparkConf conf = new SparkConf();
        conf.setMaster("local").setAppName("javaSparkWC");
        JavaSparkContext ctx = new JavaSparkContext(conf);


        simpleWC(ctx);
//        commonWC(ctx);



        ctx.stop();
    }

    private static void simpleWC(JavaSparkContext ctx) {
        ctx.textFile(fileName)
                .flatMap(x -> Arrays.asList(x.split(" ")))
                .mapToPair(x -> new Tuple2<>(x, 1))
                .reduceByKey((x, y) -> x + y)
                .foreach(x->System.out.println(x));
    }

    private static void commonWC(JavaSparkContext ctx) {
        JavaRDD<String> lines = ctx.textFile(fileName);

        JavaRDD<String> words = lines.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public Iterable<String> call(String line) throws Exception {
                return Arrays.asList(line.split(" "));
            }
        });

        JavaPairRDD<String, Integer> wordPair = words.mapToPair(new PairFunction<String, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(String word) throws Exception {
                return new Tuple2<>(word, 1);
            }
        });

        JavaPairRDD<String, Integer> result = wordPair.reduceByKey(new Function2<Integer, Integer, Integer>() {
            @Override
            public Integer call(Integer oldVal, Integer newVal) throws Exception {
                return oldVal + newVal;
            }
        });

        result.foreach(new VoidFunction<Tuple2<String, Integer>>() {
            @Override
            public void call(Tuple2<String, Integer> tuple2) throws Exception {
                System.out.println("(" + tuple2._1 + "," + tuple2._2 + ")");
            }
        });
    }


}
